File: day01/solution.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
)

func check(err error) {
	if err != nil {
		log.Fatalf("Error: %v\n", err)
	}
}

func nextElf(scanner *bufio.Scanner) (bool, uint) {
	var sum uint = 0
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			return true, sum
		} else {
			value, err := strconv.Atoi(line)
			check(err)
			sum += uint(value)
		}
	}
	return false, sum
}

func main() {
	file, err := os.Open("inputs")
	check(err)
	defer file.Close()

	// Stream the contents and collect the totals
	scanner := bufio.NewScanner(file)

	// Accumulator for each elf's total calories.
	acc := make([]uint, 4)

	for more, calories := nextElf(scanner); more; more, calories = nextElf(scanner) {
		acc = append(acc, calories)
	}

	// Sort the accumulator in descending order
	sort.Slice(acc, func(i, j int) bool {
		return acc[i] > acc[j]
	})

	// Part 1: Highest sum overall
	fmt.Printf("Highest value: %v\n", acc[0])

	// Part 2: Sum of highest 3.
	var sum uint = 0
	for _, v := range acc[0:3] {
		sum += v
	}
	fmt.Printf("Sum of top 3: %v\n", sum)
}

File: day01/solution.exs
defmodule Functions do
  def chunk_while(item, acc)
  def chunk_while("", acc), do: {:cont, acc, []}
  def chunk_while(item, acc), do: {:cont, [String.to_integer(item) | acc]}

  def chunk_end(acc)
  def chunk_end([]), do: {:cont, []}
  def chunk_end(acc), do: {:cont, acc, []}
end

# Stream each line in the file and collect the top 3 total calories.
{top, _} = File.stream!("inputs", [:utf8], :line)
  |> Stream.map(&String.trim/1)
  |> Stream.chunk_while([], &Functions.chunk_while/2, &Functions.chunk_end/1)
  |> Stream.reject(&Enum.empty?/1)
  |> Stream.map(fn values -> Enum.reduce(values, &(&1 + &2)) end)
  |> Enum.sort(:desc)
  |> Enum.split(3)

# Part 1: What is the highest total?
top
  |> Enum.at(0)
  |> IO.inspect(label: "Highest")

# Part 2: What is the sum of the 3 highest?
top
  |> Enum.reduce(0, &(&1 + &2))
  |> IO.inspect(label: "Top 3 sum")
